src/VisualStudio/Core/Def/Implementation/UnusedReferences/RemoveUnusedReferencesCommandHandler.cs

```diff
@@ -0,0 +1,319 @@
+            {
+                Owner = ownerHwnd
+            };
+
+            uiShell.CenterDialogOnWindow(windowHelper.Handle, IntPtr.Zero);
+
+            return dialog;
+        }
+
+        private void ApplyUnusedReferenceUpdates(Project project, ImmutableArray<ReferenceUpdate> referenceUpdates, CancellationToken cancellationToken)
+        {
+            ThreadHelper.JoinableTaskFactory.Run(
+                () => _unusedReferencesService.UpdateReferencesAsync(project, referenceUpdates, cancellationToken));
+        }
+
+        private static bool TryGetPropertyValue(IVsHierarchy hierarchy, string propertyName, out string? propertyValue)
+        {
+            if (hierarchy is not IVsBuildPropertyStorage storage)
+            {
+                propertyValue = null;
+                return false;
+            }
+
+            return ErrorHandler.Succeeded(storage.GetPropertyValue(propertyName, null, (uint)_PersistStorageType.PST_PROJECT_FILE, out propertyValue));
+        }
+
+        private bool TryGetSelectedProjectHierarchy([NotNullWhen(returnValue: true)] out IVsHierarchy? hierarchy)
+        {
+            hierarchy = null;
+
+            // Get the DTE service and make sure there is an open solution
+            if (_serviceProvider?.GetService(typeof(EnvDTE.DTE)) is not EnvDTE.DTE dte ||
+                dte.Solution == null)
+            {
+                return false;
+            }
+
+            var selectionHierarchy = IntPtr.Zero;
+            var selectionContainer = IntPtr.Zero;
+
+            // Get the current selection in the shell
+            if (_serviceProvider.GetService(typeof(SVsShellMonitorSelection)) is IVsMonitorSelection monitorSelection)
+            {
+                try
+                {
+                    monitorSelection.GetCurrentSelection(out selectionHierarchy, out var itemId, out var multiSelect, out selectionContainer);
+                    if (selectionHierarchy != IntPtr.Zero)
+                    {
+                        hierarchy = Marshal.GetObjectForIUnknown(selectionHierarchy) as IVsHierarchy;
+                        Debug.Assert(hierarchy != null);
+                        return hierarchy != null;
+                    }
+                }
+                catch (Exception)
+                {
+                    // If anything went wrong, just ignore it
+                }
+                finally
+                {
+                    // Make sure we release the COM pointers in any case
+                    if (selectionHierarchy != IntPtr.Zero)
+                    {
+                        Marshal.Release(selectionHierarchy);
+                    }
+
+                    if (selectionContainer != IntPtr.Zero)
+                    {
+                        Marshal.Release(selectionContainer);
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private bool IsBuildActive()
```

Comment:
Seems this duplicates existing logic: https://github.com/dotnet/roslyn/blob/7aeb78f800e9144f3485d4e9201c455e7afb6f45/src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioDiagnosticAnalyzerService.cs#L330-L351

Can we move to a shared helper or extension method on service provider?

Reply:
Moved to a shared helper.